pkg/runner/test_runner2.go

```diff
@@ -0,0 +1,845 @@
+			Targets: tr.targets,
+		}
+		resumeState, runErr = json.Marshal(rs)
+		if runErr != nil {
+			tr.log.Errorf("unable to serialize the state: %s", runErr)
+		} else {
+			runErr = statectx.ErrPaused
+		}
+	default:
+	}
+
+	return resumeState, runErr
+}
+
+func (tr *TestRunner) waitStepRunners(ctx statectx.Context) error {
+	tr.log.Debugf("waiting for step runners to finish")
+	swch := make(chan struct{})
+	go func() {
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		for {
+			ok := true
+			for _, ss := range tr.steps {
+				// numRunning == 1 is also acceptable: we allow the Run() goroutine
+				// to continue in case of error, if the result processor decided
+				// to abandon its runner, there's nothing we can do.
+				switch {
+				case !ss.stepRunning && !ss.readerRunning:
+				// Done
+				case ss.stepRunning && ss.readerRunning:
+					// Still active
+					ok = false
+				case !ss.stepRunning && ss.readerRunning:
+					// Transient state, let it finish
+					ok = false
+				case ss.stepRunning && !ss.readerRunning:
+					// This is possible if plugin got stuck and result processor gave up on it.
+					// If so, it should have left an error.
+					if ss.runErr == nil {
+						tr.log.Errorf("%s: result processor left runner with no error", ss)
+						// There's nothing we can do at this point, fall through.
+					}
+				}
+			}
+			if ok {
+				close(swch)
+				return
+			}
+			tr.cond.Wait()
+		}
+	}()
+	var err error
+	select {
+	case <-swch:
+		tr.log.Debugf("step runners finished")
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		err = tr.checkStepRunners()
+	case <-time.After(tr.shutdownTimeout):
+		tr.log.Errorf("step runners failed to shut down correctly")
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		// If there is a step with an error set, use that.
+		err = tr.checkStepRunners()
+		// If there isn't, enumerate ones that were still running at the time.
+		nrerr := &cerrors.ErrTestStepsNeverReturned{}
+		if err == nil {
+			err = nrerr
+		}
+		for _, ss := range tr.steps {
+			if ss.stepRunning {
+				nrerr.StepNames = append(nrerr.StepNames, ss.sb.TestStepLabel)
+				// We cannot make the step itself return but we can at least release the reader.
+				tr.safeCloseOutCh(ss)
+			}
+		}
+	}
+	// Emit step error events.
+	for _, ss := range tr.steps {
+		tr.log.Debugf("%s %v", ss, ss.runErr)
+		if ss.runErr != nil && ss.runErr != statectx.ErrPaused && ss.runErr != statectx.ErrCanceled {
+			if err := ss.emitEvent(EventTestError, nil, ss.runErr.Error()); err != nil {
+				tr.log.Errorf("failed to emit event: %s", err)
+			}
+		}
+	}
+	return err
+}
+
+func (tr *TestRunner) injectTarget(ctx statectx.Context, tgs *targetState, ss *stepState, log *logrus.Entry) error {
+	log.Debugf("%s: injecting into %s", tgs, ss)
+	select {
+	case ss.inCh <- tgs.tgt:
+		// Injected successfully.
+		err := ss.ev.Emit(testevent.Data{EventName: target.EventTargetIn, Target: tgs.tgt})
+		tr.mu.Lock()
+		defer tr.mu.Unlock()
+		if tgs.CurPhase == targetStepPhaseBegin {
```

Comment:
because, in fact, after injecting the target the goroutine can get descheduled and blocked for enough time for the result to be received and state to transition to targetStepPhaseResultPending.
added a comment.